fhir_zulip_nlp/fhir_zulip_nlp.py

```diff
@@ -27,117 +32,85 @@
-            {"operator": "search", "operand": key},
-            ],
-        }
-    result_new = client.get_messages(newest)
-    datenew = result_new['messages'][0]['timestamp']
-    datenew1 = datetime.utcfromtimestamp(datenew).strftime('%Y-%m-%d %H:%M:%S')
-    vals['newest time']=datenew1
-
-    oldest = {
-        "anchor": "oldest",
-        "num_before": 0,
-        "num_after": 1,
-        "narrow": [
-            {"operator": "stream", "operand": "terminology"},
-            {"operator": "search", "operand": key},
-            ],
-        }
-    result_old = client.get_messages(oldest)
-    dateold = result_old['messages'][0]['timestamp']
-    dateold1 = datetime.utcfromtimestamp(dateold).strftime('%Y-%m-%d %H:%M:%S')
-    vals['oldest time']=dateold1
-
-    return vals
-'''
+def format_timestamp(timestamp: int):
+    """Format datetime string"""
+    return datetime.utcfromtimestamp(timestamp).strftime('%Y-%m-%d %H:%M:%S')
 
 
-def messagepull(anchor: str, before: int, after: int, keyword: str) -> Dict:
+def message_pull(anchor: str, before: int, after: int, keyword: str) -> Dict:
     """Get messages"""
-    client = zulip.Client(config_file=CONFIG['zuliprc_path'])
     request = {
-            "anchor": anchor,
-            "num_before": before,
-            "num_after": after,
-            "narrow": [
-                {"operator": "stream", "operand": "terminology"},
-                {"operator": "search", "operand": keyword},
-            ],
-            }
-    result = client.get_messages(request)
+        "anchor": anchor,
+        "num_before": before,
+        "num_after": after,
+        "narrow": [
+            {"operator": "stream", "operand": "terminology"},
+            {"operator": "search", "operand": keyword}]}
+    result = CLIENT.get_messages(request)
     return result 
 
 
-def format_timestamp(i):
-    """Format datetime string"""
-    return datetime.utcfromtimestamp(i).strftime('%Y-%m-%d %H:%M:%S')
-
-
-# TODO: will probably want to pass (i) chat.fhir.org, (ii) list of streams
-def get_data(key):
+# TODO #3: for caching, we should store the datetime of the newest message for each keyword
+#  This will solve "Every user is limited to 200 total API requests per minute.":
+#  'api usage exceeded rate limit'. retry after (float) seconds.
+# TODO: use the res['retry-after'] to wait a certain amount of seconds
+def query_keyword(keyword: str):
     """Fetch data from API
     Resources
       - Get all streams (probably not needed): https://zulip.com/api/get-streams
     """
-    bob = False
     i = 1
-    allmess: List = []
+    allmess: List[Dict] = []
     while True:
-        if bob:
-            break
-        res = messagepull(str(0), 0, i, key)
-        allmess = res['messages']
+        res: Dict = message_pull(anchor=str(0), before=0, after=i, keyword=keyword)
         i += 1
-        bob = res['found_newest']
+        try:
+            allmess += res['messages']
+        except KeyError:  # rate limit
```

Comment:
@rohaher Found the solution to the rate limit! When I paused the debugger here, I saw that there was a `'retry-after'` value. Very handy.

Caching will also help w/ this, but I haven't gotten to that yet.